### ISSUE TITLE
estimateSTR.R 

### ISSUE THREAD

**Issue by holtjma** [Oshlack/STRetch#14]
Hello,

I followed the install instructions provided using R version 3.4.1.  I ran the test instructions provided on the wiki page and didn't receive any error messages.

I then followed the instructions on the "Running STRetch" wiki pages with a single WGS sample from FASTQ files using the already provided STRetch_wgs_pipeline.groovy file.  I thought the program ran to completion, but when I inspected using "bpipe errors" I saw the following error message:

```
 [output]$ ~/githubDL/STRetch/tools/bin/bpipe errors

============================== Found 1 failed commands from run 18623 ==============================

============================================ Command 11 ============================================

Command    : Rscript /path/githubDL/STRetch/scripts/estimateSTR.R  --model /path/githubDL/STRetch/scripts/STRcov.model.csv
Exit Code  : 1
Output     : 

	Error in apply(z, 1, function(x) { : dim(X) must have a positive length
	Execution halted

```

Everything up to this point appears to have run as expected (i.e. no upstream errors, a BAM file and index file for the sample is present in the directory).  I also have .locus_counts, .median_cov, and .STR_counts files, but I'm guessing they are not accurate due to the aforementioned error.

I think this is likely a bug, but if not do you have any advice on proceeding with fixing it on my end?

Thanks in advance!

**Comment by hdashnow** [comment]
Hi Matt,

It sounds like the installation was a success and all but the final stage of the pipeline ran okay. So I'm guessing there's either a bug in the final stage, or you might be using different versions of some of the R packages (which is something I may need to account for anyway). 

I also often see these types of errors when there's something unusual in the input data that I haven't encountered before, and so have provided a nice error message for.

Would you be willing to share the STRetch intermediate files with me, so I can reproduce the error on my machine? The files I would need are: 
.STR_counts  
.locus_counts  
.median_cov  

My email address is harriet.dashnow@mcri.edu.au

I assume from the pipeline you chose that you have whole genome data? Is it PCR+ or PCR-free?

**Comment by holtjma** [comment]
Actually, I think I might have figured this out (by accident of course).  I was running a single sample only, and in my `pipeline_config.groovy` file I had `CONTROL=""`.  When I uncommented the other line `CONTROL="~/githubDL/STRetch/reference-data/PCRfreeWGS.controls.tsv"` the final file was produced without error.  

My guess is just that it tried to run the statistical test on an empty control dataset and crashed as a result.  Are there any options to run STRetch without performing this calculation in the event of a singleton where I just want the calculated STR lengths?

**Comment by hdashnow** [comment]
Hi Matt,

The STR allele size estimates are only based on the reads in that sample, not the control information. So those wont change at all. The control is only used to calculate the outlier z scores and corresponding p values, asking the question: is the allele in this sample significantly longer than other samples? If you run a single sample, you simply get no values for the outlier z scores and p values. I shall add these comments to the wiki, also.

What you're seeing does sound like a bug, as it should be able to run without controls. In fact, did you have CONTROL="" when running the sample data you used to check the install? I'd still love to take a look at your intermediate files so I can nut it out.

Warm regards,
Harriet

**Comment by holtjma** [comment]
Harriet,

Okay, as I work-around I can just use the controls that are already provided and perform any recalculation later if I need to.

When I ran it the first time, I had CONTROL="".  I tried re-running with that and it crashed on me.  But like I said above, using the provided controls worked.  I'm not sure I can send the information (PHI reason, let me talk to my PI), but I figured out how to reproduce it over here with relative ease:

1. I went into a directory where I previously ran a single sample from the FASTQ files using the provided controls.  
2. I deleted the <SAMPLE>.STRs.tsv and STRs.tsv files.  Re-running "bpipe" from here will re-create the STRs files (aka, no errors and finishes in roughly 10-20 seconds).
3. I commented out the control line in the pipeline_config and uncommented the ```CONTROL=""``` line.
4.  Re-run "bpipe" command and you get the error message I mentioned earlier:
```
======================================= Stage estimate_size ========================================
Error in apply(z, 1, function(x) { : dim(X) must have a positive length
Execution halted
ERROR: Command failed with exit status = 1 : 

Rscript [PATH]/githubDL/STRetch/scripts/estimateSTR.R  --model [PATH]/githubDL/STRetch/scripts/STRcov.model.csv 
```

I _think_ there should be no issue sharing the information since it's all derived numerical values (I don't see any PHI), but I still need to clear it.  If so, I'll be in touch via email soon.

Matt

**Comment by holtjma** [comment]
I sent an email with the files you requested from the run, let me know if it wasn't received.

**Comment by ssadedin** [comment]
Just a note: I hit a similar error (not identical) to this when trying to run the WGS pipeline with the demo data without controls. The exome pipeline worked. When I provided controls to the WGS pipeline that also worked. 

My particular error occurs here:

```R
> z = apply(total_assigned_wide, 1, hubers.z)
 Show Traceback
 
 Rerun with Debug
 Error in if (s0 == 0) return(list(mu = mu0, s = 0)) : 
  missing value where TRUE/FALSE needed 
```

The value of total_assigned_wide is all `Inf`:

![image](https://user-images.githubusercontent.com/138868/29014916-eb37cd30-7b8e-11e7-989d-4692f991cf5d.png)

**Comment by hdashnow** [comment]
Fixed by Pull request #21
